### PR TITLE
task(featuresFlags): add feature flags to events

### PR DESCRIPTION
**Pull request by rich-bugsnag**
## Goal

This is Part 1 of the Unity feature flags task.

This PR adds feature flag opperations to the Configuration and Client (fallback and native)

Part 2 will add the opperations to callbacks.

## Design

Feature flags are stored in the Client. When start is called, any flags set in the config will be transfered to the client.

Any feature flag operation on the client is also passed through to the native client so that they are in sync and native errors report the correct flags. 

## Changeset

- Created the serialisable payload class "FeatureFlag"
- Created the interface "IFeatureFlagStore".
- Implemented that interface in the "Configuration", "Client" and "NativeClient" classes

## Testing

Added Full CI tests for Fallback, Cocoa and Android

### CODE REVIEW

File: src/BugsnagUnity/Native/Android/NativeInterface.cs [hunk 1]
```diff
@@ -998,5 +1007,32 @@ public LastRunInfo GetlastRunInfo()
...
         }
 
+        private IntPtr GetClientRef()
+        {
+            return CallNativeObjectMethodRef("getClient", "()Lcom/bugsnag/android/Client;", new object[] { });
+        }
+
+        public void AddFeatureFlag(string name, string varient)
+        {
+            var methodId = AndroidJNI.GetMethodID(AndroidJNI.FindClass("com/bugsnag/android/Client"), "addFeatureFlag", "(Ljava/lang/String;Ljava/lang/String;)V");
```
Comment (nickdowell) [inline src/BugsnagUnity/Native/Android/NativeInterface.cs:1017]: Is getting the JNI method and finding the class slow? Do we need to cache the methodId?

File: src/BugsnagUnity/Native/Android/NativeEvent.cs [hunk 2]
```diff
@@ -148,5 +148,6 @@ public void SetUser(string id, string email, string name)
         {
             NativePointer.Call("setUser", id, email, name);
         }
+
```
Comment (nickdowell) [inline src/BugsnagUnity/Native/Android/NativeEvent.cs:151]: Seems like an unnecessary change 😂

File: src/BugsnagUnity.m [hunk 3]
```diff
@@ -387,6 +387,27 @@ void bugsnag_setReleaseStage(const void *configuration, char *releaseStage) {
...
 
+void bugsnag_addFeatureFlagOnConfig(const void *configuration, char *name, char *varient) {
+    NSString *ns_name = name == NULL ? nil : [NSString stringWithUTF8String: name];
+    NSString *ns_varient = varient == NULL ? nil : [NSString stringWithUTF8String: varient];
+    [(__bridge BugsnagConfiguration *)configuration addFeatureFlagWithName:ns_name variant:ns_varient];
+}
+
+void bugsnag_addFeatureFlag(char *name, char *varient) {
+    NSString *ns_name = name == NULL ? nil : [NSString stringWithUTF8String: name];
+    NSString *ns_varient = varient == NULL ? nil : [NSString stringWithUTF8String: varient];
```
Comment (nickdowell) [inline src/BugsnagUnity.m:398]: It's spelled `variant`, not `varient`

File: src/BugsnagUnity/Native/Android/NativeInterface.cs (hunk 1 above)
Comment (lemnik) [reply src/BugsnagUnity/Native/Android/NativeInterface.cs:1017]: These looks aren't cheap, but there's also a question on how frequently they are called (up-front caching might not be worth it)?

That being said: caching them up-front also keeps things consistent with the other functions we call?

Comment (rich-bugsnag) [reply src/BugsnagUnity/Native/Android/NativeInterface.cs:1017]: Cached 👍

File: src/BugsnagUnity.m (hunk 3 above)
Comment (rich-bugsnag) [reply src/BugsnagUnity.m:398]: Fixed

File: features/fixtures/maze_runner/Assets/Scripts/Main.cs [hunk 4]
```diff
@@ -151,6 +151,16 @@ void PrepareConfigForScenario(Configuration config, string scenario)
...
         switch (scenario)
         {
+            case "FeatureFlagsConfigClearAll":
+                config.AddFeatureFlag("testName1", "testVarient1");
+                config.AddFeatureFlag("testName2", "testVarient2");
+                config.ClearFeatureFlags();
+                break;
+            case "FeatureFlagsInConfig":
+                config.AddFeatureFlag("testName1", "testVarient1");
+                config.AddFeatureFlag("testName2", "testVarient2");
```
Comment (lemnik) [inline features/fixtures/maze_runner/Assets/Scripts/Main.cs:161]:
```suggestion
            case "FeatureFlagsConfigClearAll":
                config.AddFeatureFlag("testName1", "testVariant1");
                config.AddFeatureFlag("testName2", "testVariant2");
                config.ClearFeatureFlags();
                break;
            case "FeatureFlagsInConfig":
                config.AddFeatureFlag("testName1", "testVariant1");
                config.AddFeatureFlag("testName2", "testVariant2");
```

File: features/fixtures/maze_runner/Assets/Scripts/Main.cs [hunk 5]
```diff
@@ -360,6 +370,22 @@ void RunScenario(string scenario)
...
+                throw new Exception("FeatureFlags");
+                break;
+            case "FeatureFlagsConfigClearAll":
+                throw new Exception("FeatureFlags");
+                break;
+            case "FeatureFlagsInConfig":
+                throw new Exception("FeatureFlags");
+            case "FeatureFlagsAfterInit":
+                Bugsnag.AddFeatureFlag("testName1", "testVarient1");
+                Bugsnag.AddFeatureFlag("testName2", "testVarient2");
```
Comment (lemnik) [inline features/fixtures/maze_runner/Assets/Scripts/Main.cs:386]:
```suggestion
                Bugsnag.AddFeatureFlag("testName1", "testVariant1");
                Bugsnag.AddFeatureFlag("testName2", "testVariant2");
                Bugsnag.ClearFeatureFlags();
                throw new Exception("FeatureFlags");
                break;
            case "FeatureFlagsConfigClearAll":
                throw new Exception("FeatureFlags");
                break;
            case "FeatureFlagsInConfig":
                throw new Exception("FeatureFlags");
            case "FeatureFlagsAfterInit":
                Bugsnag.AddFeatureFlag("testName1", "testVariant1");
                Bugsnag.AddFeatureFlag("testName2", "testVariant2");
```

File: features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs [hunk 6]
```diff
@@ -135,6 +140,11 @@ private Configuration PreapareConfigForScenario(string scenarioName)
 
         switch (scenarioName)
         {
+            case "Feature Flags In Config":
+                config.AddFeatureFlag("testName1","testVarient1");
+                config.AddFeatureFlag("testName2", "testVarient2");
```
Comment (lemnik) [inline features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs:145]:
```suggestion
                config.AddFeatureFlag("testName1","testVariant1");
                config.AddFeatureFlag("testName2", "testVariant2");
```

File: src/BugsnagUnity/Native/Android/NativeEvent.cs (hunk 2 above)
Comment (rich-bugsnag) [reply src/BugsnagUnity/Native/Android/NativeEvent.cs]: I was going to do callbacks in the same PR but changed my mind and removed some code that was previously checked in

File: features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs [hunk 7]
```diff
@@ -368,7 +382,33 @@ public void DoTestAction(string scenarioName)
     {
         switch (scenarioName)
         {
-
+            case "Feature Flags After Init Clear All":
+                Bugsnag.AddFeatureFlag("testName1", "testVarient1");
+                Bugsnag.AddFeatureFlag("testName2", "testVarient2");
```
Comment (lemnik) [inline features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs:387]:
```suggestion
                Bugsnag.AddFeatureFlag("testName1", "testVariant1");
                Bugsnag.AddFeatureFlag("testName2", "testVariant2");
```

File: features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs [hunk 8]
```diff
@@ -368,7 +382,33 @@ public void DoTestAction(string scenarioName)
...
+                Bugsnag.ClearFeatureFlags();
+#if UNITY_ANDROID
+                MobileNative.TriggerBackgroundJavaCrash();
+#elif UNITY_IOS
+                NativeException();
+#endif
+                break;
+            case "Feature Flags After Init":
+                Bugsnag.AddFeatureFlag("testName1", "testVarient1");
+                Bugsnag.AddFeatureFlag("testName2", "testVarient2");
```
Comment (lemnik) [inline features/fixtures/maze_runner/Assets/Scripts/MobileScenarioRunner.cs:397]:
```suggestion
                Bugsnag.AddFeatureFlag("testName1", "testVariant1");
                Bugsnag.AddFeatureFlag("testName2", "testVariant2");
```

File: features/android/android_feature_flags.feature [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        And I wait for 2 seconds
+        When I clear any error dialogue
+        And I relaunch the Unity mobile app
+        When I run the "Start SDK" mobile scenario
+        Then I wait to receive an error
+        
+        And the exception "errorClass" equals "java.lang.RuntimeException"
+
+        And the event "featureFlags.0.featureFlag" equals "testName2"
+        And the event "featureFlags.0.variant" equals "testVarient2"
```
Comment (lemnik) [inline features/android/android_feature_flags.feature:17]:
```suggestion
        And the event "featureFlags.0.variant" equals "testVariant2"
```

File: features/android/android_feature_flags.feature [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
```
Comment (lemnik) [inline features/android/android_feature_flags.feature:31]:
```suggestion
        And the event "featureFlags.0.variant" equals "testVariant2"
```

File: features/ios/ios_feature_flags.feature [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        When I run the "Feature Flags In Config" mobile scenario
+        And I wait for 2 seconds
+        And I relaunch the Unity mobile app
+        When I run the "Start SDK" mobile scenario
+        Then I wait to receive an error
+        
+        And the exception "errorClass" equals "St13runtime_error"
+
+        And the event "featureFlags.0.featureFlag" equals "testName2"
+        And the event "featureFlags.0.variant" equals "testVarient2"
```
Comment (lemnik) [inline features/ios/ios_feature_flags.feature:17]:
```suggestion
        And the event "featureFlags.0.variant" equals "testVariant2"
```

File: features/ios/ios_feature_flags.feature [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        When I run the "Feature Flags After Init" mobile scenario
+        And I wait for 2 seconds
+        And I relaunch the Unity mobile app
+        When I run the "Start SDK" mobile scenario
+        Then I wait to receive an error
+        
+        And the exception "errorClass" equals "St13runtime_error"
+
+        And the event "featureFlags.0.featureFlag" equals "testName2"
+        And the event "featureFlags.0.variant" equals "testVarient2"
```
Comment (lemnik) [inline features/ios/ios_feature_flags.feature:30]:
```suggestion
        And the event "featureFlags.0.variant" equals "testVariant2"
```

**Review by lemnik — APPROVED** [review]
LGTM Pending some suggestions

**Review by nickdowell — APPROVED** [review]
